### PR TITLE
[BUG] Two phase pause to prevent deadlock

### PR DESCRIPTION
**Pull request by hao-aaron**
### Motivation

With data-parallel (DP) engines, returning from pause does not guarantee that engines stay paused. After `pause_scheduler` returns, a stale `START_DP_WAVE` message can re-activate an engine that was supposed to be idle. If a caller then performs a collective operation (e.g. weight loading, KV cache management) expecting all engines to be quiesced, only some ranks participate — resulting in a deadlock on the collective.

A simpler fix — unconditionally ignoring `START_DP_WAVE` while paused — was considered but rejected. If one rank pauses while another still has in-flight work, the working rank will eventually need the paused rank to participate in the `_has_global_unfinished_reqs` all-reduce to finish its wave. Ignoring the wave message means the paused rank never enters the stepping loop, so the all-reduce never completes.

### Solution

The core idea is to split pause into two phases. In the first phase, each engine pauses its scheduler locally but continues to step and honor `START_DP_WAVE` messages — this keeps it available for the all-reduce so that other ranks can still finish their work. In the second phase, once every rank has signaled that it wants to pause, they all agree (via the existing periodic all-reduce) to stop together and begin ignoring `START_DP_WAVE`. Because the decision to ignore wave messages is collective, no rank can be left behind, and no stale message can re-wake an engine after the group has agreed to pause. Resume clears both flags, restoring normal behavior.

### Changes

**Two-phase pause protocol** (`DPEngineCoreProc.pause_scheduler`):

- **Phase 1 — local intent:** When a DP engine receives a pause request, it pauses the scheduler as before (behavior unchanged) and sets a `pending_pause` flag. During this phase, `START_DP_WAVE` messages are still respected — the engine keeps stepping (including dummy batches if idle) so it can participate in the all-reduce. If the engines are already idle, it kick-starts them by sending a `start_wave` so all ranks enter the stepping loop.
- **Phase 2 — consensus:** During the periodic all-reduce (every 32 steps), engines now synchronize both "has unfinished work" and "wants to pause" in a single fused operation (`sync_dp_state`). Once **all** ranks report `pending_pause`, they collectively stop and set `ignore_start_dp_wave` — from this point on, stale wave messages are ignored, guaranteeing the engines stay quiesced.

**Combined all-reduce** (`ParallelConfig.sync_dp_state`):

Replaces the old `has_unfinished_dp` with a single 2-element SUM all-reduce:
- `[0]`: logical OR — any rank has unfinished work (or a pending pause keeping it alive).
- `[1]`: logical AND via `SUM == dp_size` — all ranks want to pause.

No additional communication cost; the existing all-reduce is reused with a wider tensor.

### Considerations

**Fast sync path (every-32-steps optimization):** The all-reduce only runs every 32 forward steps. Between checkpoints, a pausing engine signals "unfinished" (`has_unfinished=local_unfinished or self.pending_pause`) so it continues stepping through the gap. This means pause latency is bounded by up to 32 dummy steps — negligible in practice, and avoids adding a second synchronization point on the hot path.

**Future optimization:** When `pending_pause` is set, we could bypass the 32-step interval and trigger the all-reduce immediately (or after a much smaller interval). This would reduce pause latency from up to 32 dummy steps to nearly instant consensus, at the cost of a slightly more complex check in the hot path.

-------------------------------------
thanks to @junjzhang for identifying the issue https://github.com/vllm-project/vllm/issues/36594, and helping to test out and debug this solution!

Co-authored-by: Junjie Zhang <[junj.jay.zhang@gmail.com](mailto:junj.jay.zhang@gmail.com)>

### CODE REVIEW

File: vllm/v1/engine/core.py [hunk 1]
```diff
@@ -1759,7 +1818,20 @@ def _has_global_unfinished_reqs(self, local_unfinished: bool) -> bool:
         if self.step_counter % 32 != 0:
             return True
```
[the remark this answers is not in the document]
Comment (hao-aaron) [reply vllm/v1/engine/core.py:1850]: this is intended behavior, for example pause mode finish needs to continue running.

**Comment by junjzhang** [comment]
Tested this on an EP MoE setup (gpt-oss-120b, DP=2 / TP=8 / EP=16). The two-phase pause protocol itself works correctly. But the change to `DPEngineCoreProc.add_request` introduces a regression that hangs the DP all-reduce during normal inference:

```diff
             elif (
                 not self.engines_running
                 and self.scheduler.pause_state == PauseState.UNPAUSED
             ):
-                self.engines_running = True
                 self.output_queue.put_nowait(
                     (-1, EngineCoreOutputs(start_wave=self.current_wave))
                 )
```

When a stale-wave request lands on an idle engine, this engine sends `start_wave` through the coordinator, expecting the `START_DP_WAVE` broadcast to wake everyone. But `_handle_client_request` excludes the sender via `exclude_eng_index`:

```python
if exclude_eng_index != self.engine_index and new_wave >= self.current_wave:
    ...
    if not self.engines_running:
        self.engines_running = True
```

So the sender is excluded from its own broadcast and never wakes. The peer engine wakes, enters the stepping loop, and every forward step does the per-step DP `num_tokens_across_dp_cpu` all-reduce — which blocks forever waiting for the non-participating sender. 600s watchdog timeout, NCCL crash.

Fix is to restore the local wake-up — the pre-PR behavior was already correct given `exclude_eng_index`:

```python
elif not self.engines_running and self.scheduler.pause_state == PauseState.UNPAUSED:
    self.engines_running = True
    self.output_queue.put_nowait((-1, EngineCoreOutputs(start_wave=self.current_wave)))
```

With this one-line addition on top of your PR, our workload runs through multiple weight sync rounds cleanly. Otherwise every run crashes.

**Comment by hao-aaron** [comment]
thanks for the validation @junjzhang!

File: vllm/config/parallel.py [hunk 2]
```diff
@@ -663,6 +650,37 @@ def has_unfinished_dp(dp_group: ProcessGroup, has_unfinished: bool) -> bool:
...
+        has_unfinished_global is true if any rank has unfinished work,
+        or if some ranks are waiting for a pause consensus.
+
+        Returns:
+            (has_unfinished_global, pause_consensus)
+        """
+        tensor = torch.tensor(
+            [int(has_unfinished), int(pending_pause)],
+            dtype=torch.int32,
+            device="cpu",
```
Comment (njhill) [inline vllm/config/parallel.py]:
```suggestion
            [int(has_unfinished), int(pending_pause)], dtype=torch.int32, device="cpu"
```

File: vllm/v1/engine/core.py [hunk 3]
```diff
@@ -1677,36 +1703,60 @@ def add_request(self, request: Request, request_wave: int = 0):
...
             if exclude_eng_index != self.engine_index and (
                 new_wave >= self.current_wave
             ):
                 self.current_wave = new_wave
                 if not self.engines_running:
-                    logger.debug("EngineCore starting idle loop for wave %d.", new_wave)
+                    logger.debug(
+                        "EngineCore starting idle loop for wave %d.",
+                        new_wave,
+                    )
```
Comment (njhill) [inline vllm/v1/engine/core.py:1772]: revert?

File: vllm/config/parallel.py [hunk 4]
```diff
@@ -663,6 +650,37 @@ def has_unfinished_dp(dp_group: ProcessGroup, has_unfinished: bool) -> bool:
         aggregated_has_unfinished = bool(tensor.item())
         return aggregated_has_unfinished
 
+    @staticmethod
+    def sync_dp_state(
+        dp_group: ProcessGroup,
+        has_unfinished: bool,
+        pending_pause: bool,
+        dp_size: int,
```
Comment (njhill) [inline vllm/config/parallel.py]:
```suggestion
        dp_group: ProcessGroup, has_unfinished: bool, pending_pause: bool
```

Can use `dp_group.size()` instead of passing `dp_size`

File: vllm/v1/engine/core.py [hunk 5]
```diff
@@ -1786,7 +1836,19 @@ def _has_global_unfinished_reqs(self, local_unfinished: bool) -> bool:
         if self.step_counter % 32 != 0:
             return True
 
-        return ParallelConfig.has_unfinished_dp(self.dp_group, local_unfinished)
+        has_unfinished, pause_consensus = ParallelConfig.sync_dp_state(
+            self.dp_group,
+            has_unfinished=local_unfinished,
+            pending_pause=self.pending_pause,
+            dp_size=self.dp_size,
```
Comment (njhill) [inline vllm/v1/engine/core.py]:
```suggestion
            self.dp_group, has_unfinished=local_unfinished, pending_pause=self.pending_pause
```

File: vllm/v1/engine/core.py [hunk 6]
```diff
@@ -1668,6 +1650,50 @@ def shutdown(self):
...
+        confirms that **all** ranks have ``pending_pause`` set, collectively
+        stop stepping and set ``ignore_start_dp_wave`` so that stale
+        ``START_DP_WAVE`` messages cannot re-wake any engine.
+        """
+        if mode not in ("keep", "abort", "wait"):
+            raise ValueError(f"Invalid pause mode: {mode}")
+
+        def engine_idle_callback(
+            engine: "DPEngineCoreProc", future: Future[Any]
+        ) -> None:
```
Comment (njhill) [inline vllm/v1/engine/core.py]:
```suggestion
        def engine_idle_callback(engine: "DPEngineCoreProc", future: Future[Any]) -> None:
```

File: vllm/v1/engine/core.py [hunk 7]
```diff
@@ -1668,6 +1650,50 @@ def shutdown(self):
         if dp_group := getattr(self, "dp_group", None):
             stateless_destroy_torch_distributed_process_group(dp_group)
 
+    def pause_scheduler(
```
Comment (njhill) [inline vllm/v1/engine/core.py]: Since this method is almost identical to the superclass method, we could consider extracting and just overriding the part that's different.

i.e. a `def _do_pause() -> bool` which returns whether pause has completed or still pending.

Then here we would just have:

```python
def do_pause() -> bool:
        self.pending_pause = True

        # Kick-start this engine into the stepping loop so it can
        # reach the all-reduce consensus checkpoint.
        self.engines_running = True

        return False
```

and could still include the docstring from below explaining the two-phase process

**Review by njhill — COMMENTED** [review]
This looks great, thanks @hao-aaron, nice idea. Only minor comments. Apologies again for the review delay.

**Review by njhill — APPROVED** [review]
